backend/app/core/langfuse/langfuse.py

```diff
@@ -107,3 +109,112 @@ def log_error(self, error_message: str, response_id: Optional[str] = None):
 
     def flush(self):
         self.langfuse.flush()
+
+
+def observe_llm_execution(
+    session_id: str | None = None,
+    credentials: dict | None = None,
+):
+    """Decorator to add Langfuse observability to LLM provider execute methods.
+
+    Args:
+        credentials: Langfuse credentials with public_key, secret_key, and host
+        session_id: Session ID for grouping traces (conversation_id)
+
+    Usage:
+        decorated_execute = observe_llm_execution(
+            credentials=langfuse_creds,
+            session_id=conversation_id
+        )(provider_instance.execute)
+    """
+
+    def decorator(func: Callable) -> Callable:
+        @wraps(func)
+        def wrapper(completion_config: CompletionConfig, query: QueryParams, **kwargs):
+            # Skip observability if no credentials provided
+            if not credentials or not all(
+                key in credentials for key in ["public_key", "secret_key", "host"]
+            ):
```

Comment:
do we need this extensive check as if credentials are there then `"public_key", "secret_key", "host"` should be there else user can't save credentials